SETUP/db_schema.sql

```diff
@@ -118,6 +118,20 @@ CREATE TABLE `job_logs` (
   `comments` varchar(255) default NULL
 );
 
+--
+-- Table structure for table `latest_snapshot_tallies`
+--
+
+CREATE TABLE `latest_snapshot_tallies` (
+  `tally_name` char(2) NOT NULL default '',
+  `holder_type` char(1) NOT NULL default '',
+  `holder_id` int(6) unsigned NOT NULL default '0',
+  `timestamp` int(10) unsigned NOT NULL default '0',
+  `tally_delta` int(8) NOT NULL default '0',
+  `tally_value` int(8) NOT NULL default '0',
+  PRIMARY KEY (`tally_name`,`holder_type`,`holder_id`)
```

Comment:
I thought about that but I don't like the idea of making `current_tallies` not contain, well, just the current tallies. Then it becomes `current_and_last_snapshot_tallies` and we start muddling things together.

I suspect that taking snapshots is going to be faster since we're no longer relying on `past_tallies` to capture them at all -- we only use `current_tallies` and `last_snapshot_tallies`.

Reply:
> I thought about that but I don't like the idea of making `current_tallies` not contain, well, just the current tallies. Then it becomes `current_and_last_snapshot_tallies`

Personally, I think "current" is vague enough to include yesterday, but we could call it something else, like `recent_tallies`. 

> and we start muddling things together.

I think it's putting very closely related things together.

> I suspect that taking snapshots is going to be faster since we're no longer relying on `past_tallies` to capture them at all -- we only use `current_tallies` and `last_snapshot_tallies`.

I'm not sure it's going to be faster. It might even be slower, since we'd be replacing:
> for each holder on the tallyboard, insert into past_tallies
with:
> for each holder on the tallyboard, insert into latest_snapshot_tallies; then for each holder with a non-zero delta, insert into past_tallies

so it's strictly more insert commands, right? *Maybe* inserting into `latest_snapshot_tallies` will be faster than inserting into `past_tallies`, but I suspect the per-command overhead will dominate when we're doing it for thousands of users.

Not that that's an objection to this PR: having a sparse `past_tallies` is important, and a longer snapshot job would be an acceptable price to pay. *But* if we have the possibility of reducing the period of capture-uncertainty from 15-20 minutes down to (I don't know) a few seconds, I think we should look into it.

Reply:
> I'm not sure it's going to be faster. It might even be slower, since we'd be replacing:
> 
> > for each holder on the tallyboard, insert into past_tallies
> 
> with:
> 
> > for each holder on the tallyboard, insert into latest_snapshot_tallies; then for each holder with a non-zero delta, insert into past_tallies
> 
> so it's strictly more insert commands, right?

You're eliding over the fact that we're no longer querying `past_tallies` to do the snapshot -- we're now query a primary key in a small table. The original query is non-negligible because while the query is indexed-based, the index is _huge_ because the table is huge and we have such broad indexes on it so its doing a lot of disk I/O just to load the index to do that query.

For most users we're not inserting into `latest_snapshot_tallies`, we're effectively updating a row based entirely on the value of the primary key -- so the DB doesn't have to allocate more space or manage indexes. We're inserting about ~1% of the current number of rows into `past_tallies` than we're doing now so we're likely doing 1% more update commands but 99% of them are more efficient.

I'll give the "throw everything into `current_tallies`" more thought but I suspect we're going to see pretty massive performance improvements even with the proposed design because we're taking `past_tallies` out of the picture as much as possible.

And note that structurally we wouldn't want to update all tallyboards in a single command as that would violate a couple of layers of abstraction -- we'd need to do it for a tallyboard at a time.

Reply:
I've merged `last_snapshot_tallies` into `current_tallies`. I want to keep each Tallyboard snapshot self-contained so I've kept the last snapshot timestamp as a column in `current_tallies`.

Reply:
Something to note about doing the update with an `UPDATE` over many rows is that it will lock those rows for as long as it takes to do the update. Presumably that is on the order of a second or two -- the table is simple and only has 100k rows -- but anything that tries to update during that window (when someone proofreads a page) that will block until the update completes.

Reply:
I copied the `current_tallies` table from PROD to a new database on TEST and ran the UPDATE over tally_name=P1 and holder_type=U (the largest single Tallyboard) and it took <3 seconds:
```
Query OK, 67500 rows affected (2.33 sec)
Rows matched: 67827  Changed: 67500  Warnings: 0
```

TEST is slower than PROD so we should be good to go here.

Reply:
> You're eliding over the fact that we're no longer querying `past_tallies` to do the snapshot

Yup, I left that out because I thought it was negligible because of the index, but apparently not. (At least, not for the (P1,U) tallyboard, for the others it's pretty quick). (Once again, MySQL defies my intuitions.)

> And note that structurally we wouldn't want to update all tallyboards in a single command as that would violate a couple of layers of abstraction -- we'd need to do it for a tallyboard at a time.

True, it would disrepect the TallyBoard abstraction. (Not sure what other layer you're thinking of.)

So I was going to suggest a per-TallyBoard version of the 'capture everyone at once' command, but it looks like you've already got that.


> Something to note about doing the update with an `UPDATE` over many rows is that it will lock those rows for as long as it takes to do the update. Presumably that is on the order of a second or two -- the table is simple and only has 100k rows -- but anything that tries to update during that window (when someone proofreads a page) that will block until the update completes.

Right, so if someone saves-as-done (or reopens) a page in that small window, the site's response will be a bit slower than usual. But I'm guessing people won't find it too anomalous.


> I copied the `current_tallies` table from PROD to a new database on TEST and ran the UPDATE over tally_name=P1 and holder_type=U (the largest single Tallyboard) and it took <3 seconds:
> 
> TEST is slower than PROD so we should be good to go here.

Yay!